### ISSUE TITLE
Freezing a `Struct` containing a sub-nonhashable

### ISSUE THREAD

**Issue by goodboy** [jcrist/msgspec#139]
I think I've more or less had the exact same issue with `frozen=True` in `pydantic` as well but this is the issue:

```python
[ins] In [1]: import msgspec

[ins] In [2]: class Msg(msgspec.Struct, frozen=True):
         ...:     l: list[str]
         ...:

[ins] In [3]: msg = Msg([str(i) for i in range(10)])

[ins] In [4]: {msg: 10}
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 {msg: 10}

TypeError: unhashable type: 'list'
```

I realize that there might not be an easy away around this (without some kind of nested hashable obj scanner) but I figured I'd report since there's [no mention of this restriction in the docs](https://jcristharif.com/msgspec/structs.html#frozen-instances).

**Comment by goodboy** [comment]
Oh, and changing `Msg([str(i) for i in range(10)])` -> `Msg((str(i) for i in range(10)))` obviously makes it work.

**Comment by jcrist** [comment]
Thanks for raising this issue. As you've noted, there's not much msgspec can do here, this is mostly just a docs issue. I've pushed up a PR (https://github.com/jcrist/msgspec/pull/141) trying to clarify this behavior - can you take a look at that PR and see if you think that's clear and sufficient?

**Comment by goodboy** [comment]
Yup I think that PR should close this given there's no easy workaround without nasty hackery 😂

**Comment by goodboy** [comment]
@jcrist actually, thinking more about this.. is there any reason (especially since we can assume the struct becomes immutable) to not offer some kind of hash function that makes even structs with mutable embedded types unique?

In theory for a given value of some immutable thing (i realize the sub-attrs are still mutable) you could generate a unique has output no?

I'm obviously no CS expert in this area, but even if that isn't possible maybe it should be an error to define a `Struct` with mutable subtypes as attrs and use `frozen=True`?

I mean, what is the point of freezing if it doesn't work because of the embedded fields 😂

**Comment by goodboy** [comment]
Anyway yah the more I'm playing with this I see it's python that is the limitation so probably not worth the effort.

I guess it seems to me there's a difference between "immutable" and "hashable" so i wonder if maybe not assuming they're the same thing is also another way to go?

```python
class Point(msgspec.Struct, tag=True, hashable=True):
    x: int
    y: list[int]
```

So like this kinda def should throw an error, whereas frozen would just mean you can't change the field references once set?

**Comment by jcrist** [comment]
> So like this kinda def should throw an error, whereas frozen would just mean you can't change the field references once set?

For complicated reasons relating to how annotations and class definitions interplay (as well as how hashability is defined), we can't (easily) provide an error message saying that the types used in the struct aren't hashable so `hashable=True` is invalid.

We try to provide a dataclass-like api, so `frozen=True` providing both immutability and hashability makes sense here (I personally dislike the kwarg name `frozen`, but that's what cpython uses so that's what we're using). A documentation-only fix is the best we can do IMO.
